modules/programs/firefox.nix

```diff
@@ -444,6 +532,124 @@ in {
+                        definedAliases = "_definedAliases";
+                        urls = "_urls";
+                      }.${name} or name;
+
+                      inherit value;
+                    }) ((removeAttrs config [ "icon" ])
+                      // (optionalAttrs (!isAppProvided)
+                        (optionalAttrs (config ? iconUpdateURL) {
+                          # Convenience to default iconURL to iconUpdateURL so
+                          # the icon is immediately downloaded from the URL
+                          iconURL = config.iconURL or config.iconUpdateURL;
+                        } // optionalAttrs (config ? icon) {
+                          # Convenience to specify absolute path to icon
+                          iconURL = "file://${config.icon}";
+                        } // {
+                          # Required for custom engine configurations,
+                          # loadPaths are unique identifiers that are generally
+                          # formatted like: [source]/path/to/engine.xml
+                          loadPath = ''
+                            [home-manager]/programs.firefox.profiles.${profile.name}.search.engines."${
+                              replaceChars [ "\\" ] [ "\\\\" ] name
+                            }"'';
+                        })) // {
+                          # Required fields for all engine configurations
+                          inherit name isAppProvided metaData;
+                        })) orderedEngineList;
+                in engines;
+
+                metaData = optionalAttrs (profile.search.default != null) {
+                  current = profile.search.default;
+
+                  # Hash algorithm from
+                  # https://searchfox.org/mozilla-central/rev/de15f9c109f9c474d00faf8032f559c236067c06/toolkit/components/search/SearchUtils.jsm#312-334
+                  hash = let
+                    # home-manager doesn't circumvent user consent and isn't
+                    # acting maliciously. We're modifying the search outside of
+                    # Firefox, but a claim by Mozilla to remove this would be very
+                    # anti-user, and is unlikely to be an issue for our use case.
+                    disclaimer =
+                      "By modifying this file, I agree that I am doing so "
+                      + "only within $appName itself, using official, user-driven search "
+                      + "engine selection processes, and in a way which does not circumvent "
+                      + "user consent. I acknowledge that any attempt to change this file "
+                      + "from outside of $appName is a malicious act, and will be responded "
+                      + "to accordingly.";
+
+                    salt = profile.path + profile.search.default
+                      + (builtins.replaceStrings [ "$appName" ] [ "Firefox" ]
+                        disclaimer);
```

Comment:
Probably better to use a function and antiquotation rather than translating the JS code 1:1.